### ISSUE TITLE
Assertion failure with Buffer in the bytecode debug runtime

### ISSUE THREAD

**Issue by shym** [ocaml/ocaml#12103]
When running the following `bufferstress.ml` program testing `Buffer` in the bytecode debug runtime (tried it on amd64 5.0.0, but also amd64 trunk (updated the 9th of March), and amd64 5.0.0 32-bit only) I get an assertion failed:
```
$ ocamlc -g -o bufferstress.bc bufferstress.ml
$ ocamlrund bufferstress.bc
### OCaml runtime: debug mode ###
[00] file runtime/str.c; line 44 ### Assertion failed: Byte (s, temp - Byte (s, temp)) == 0
```
The code is mischievous because it uses `Buffer` without synchronization, but still…
```
let buf = Buffer.create 8

let tester () =
  Buffer.reset buf;
  Buffer.add_string buf "aaaaaa";

  let wait = Atomic.make true in
  let dom1 =
    Domain.spawn (fun () ->
        while Atomic.get wait do Domain.cpu_relax () done;
        Buffer.clear buf)
  in
  let dom2 =
    Domain.spawn (fun () ->
        Atomic.set wait false;
        Buffer.add_string buf "aaaa")
  in
  Domain.join dom1;
  Domain.join dom2

let _ =
  for _i = 0 to 1000 do
    tester ()
  done
```

(Context: I ran into this failure trying to reproduce and simplify a `Buffer` test in our [multicoretests](https://github.com/ocaml-multicore/multicoretests/) suite because we’ve observed a (reproducible) segfault in a 5.0.0 32-bit bytecode run: [tracking issue](https://github.com/ocaml-multicore/multicoretests/issues/306))

**Comment by gasche** [comment]
Can you easily try the following patch?

```diff
--- a/stdlib/buffer.ml
+++ b/stdlib/buffer.ml
@@ -166,7 +166,7 @@ let add_string b s =
   let len = String.length s in
   let position = b.position in
   let {buffer; length} = b.inner in
-  let new_position = b.position + len in
+  let new_position = position + len in
   if new_position > length then (
     resize b len;
     Bytes.blit_string s 0 b.inner.buffer b.position len;
```

**Comment by shym** [comment]
I cannot reproduce the assertion failure with that patch (tested on amd64 5.0.0 and 32-bit-only trunk)!
(I’ll try to test if I can reproduce the SEGV with this patch.)

**Comment by gasche** [comment]
Thanks! I submitted a PR at #12104.

**Comment by gasche** [comment]
@shym I wrote "closes #12103" without actually waiting for a definitive report on your end on whether the other issue is gone, which is a bit cavalier, sorry! There may remain correctness issues with Buffer[^1] but the PR was a fix for a real issue, so it is worth having in any case.

[^1]: Once you use `unsafe_{get,set}` you have trouble sleeping at night.

**Comment by shym** [comment]
Don't worry, that's perfectly fine.
Congratulations and thanks for the quick fix!
